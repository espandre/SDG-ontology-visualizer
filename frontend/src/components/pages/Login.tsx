import React, { useState } from 'react';
import {
  Heading,
  Stack,
  Flex,
  Input,
  Button,
  InputGroup,
  InputRightElement,
  Alert,
  AlertIcon,
  AlertTitle,
  AlertDescription,
  Spacer,
  FormControl,
  FormLabel,
} from '@chakra-ui/react';

import { useHistory } from 'react-router-dom';

import reducer from '../../state/store';
import { loginSuccess, loginFailed } from '../../state/reducers/loginReducer';

import { login } from '../../api/auth';

const Login: React.FC = () => {
  const history = useHistory();

  const [showPassword, setShowPassword] = useState<boolean>(false);
  const [username, setUsername] = useState<string>('');
  const [password, setPassword] = useState<string>('');

  const [missingUsername, setMissingUsername] = useState<boolean>(false);
  const [missingPassword, setMissingPassword] = useState<boolean>(false);
  const [loginError, setLoginError] = useState<boolean>(false);

  const onSubmit = async () => {
    const error: boolean = username === '' || password === '';
    setMissingUsername(username === '');
    setMissingPassword(password === '');

    if (error) return;

    const data = await login(username, password);
    if (data) {
      setLoginError(false);
      reducer.dispatch(loginSuccess(data));
      history.goBack();
    } else {
      setLoginError(true);
      reducer.dispatch(loginFailed());
    }
  };

  return (
    <Stack spacing="10">
      <Flex
        align="center"
        justify="center"
        justifyContent="space-evenly"
        h="150px"
        spacing="10"
        bg="cyan.700"
      >
        <Stack spacing="10">
          <Heading size="lg" color="white">
            Please log in
          </Heading>
        </Stack>
      </Flex>
      <Flex align="center" justify="center" justifyContent="space-evenly" spacing="10">
        <Stack
          bg="white"
          w="800px"
          align="center"
          justify="center"
          justifyContent="center"
          p="10"
          spacing="10"
          alignItems="center"
        >
          <Heading size="md" mt="5">
            The functionality you want to access requires valid credentials.
          </Heading>
          <Stack w="100%" p="10">
            {(missingUsername || missingPassword) && (
              <Alert status="error">
                <AlertIcon />
                <AlertTitle>Missing required fields:</AlertTitle>
                <AlertDescription>
                  {`${missingUsername ? 'Username' : ''}${
                    missingUsername && missingPassword ? ', ' : ''
                  }${missingPassword ? 'Password' : ''}`}
                </AlertDescription>
              </Alert>
            )}
            {loginError && (
              <Alert status="error">
                <AlertIcon />
                <AlertTitle>Error!</AlertTitle>
                <AlertDescription>Invalid username or password</AlertDescription>
              </Alert>
            )}
            <FormControl id="username" isRequired>
              <FormLabel>Username:</FormLabel>
              <Input
                onChange={(evt) => setUsername(evt.currentTarget.value)}
                errorBorderColor="crimson"
                isInvalid={missingUsername}
              />
            </FormControl>
            <FormControl id="password" isRequired>
              <FormLabel>Password:</FormLabel>
              <InputGroup size="md">
                <Input
                  type={showPassword ? 'text' : 'password'}
                  placeholder="Enter password"
                  onChange={(evt) => setPassword(evt.currentTarget.value)}
                  errorBorderColor="crimson"
                  isInvalid={missingPassword}
                />
                <InputRightElement w="4rem">
                  <Button onClick={() => setShowPassword(!showPassword)} size="sm" mr="1" w="4rem">
                    {showPassword ? 'Hide' : 'Show'}
                  </Button>
                </InputRightElement>
              </InputGroup>
            </FormControl>
            <Spacer m="2rem" />
            <Button onClick={onSubmit}>Log in</Button>
          </Stack>
        </Stack>
      </Flex>
    </Stack>
  );
};

export default Login;
