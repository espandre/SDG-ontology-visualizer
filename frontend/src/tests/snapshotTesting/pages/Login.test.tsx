import React from 'react';
import renderer from 'react-test-renderer';

import Login from '../../../components/pages/Login';

it('renders login page correctly', () => {
  const tree = renderer.create(<Login />).toJSON();
  expect(tree).toMatchSnapshot();
});
