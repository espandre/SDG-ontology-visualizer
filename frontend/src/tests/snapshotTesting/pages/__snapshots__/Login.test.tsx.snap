// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders login page correctly 1`] = `
<div
  className="chakra-stack css-efzghl"
>
  <div
    className="css-ixqi0y"
    spacing="10"
  >
    <div
      className="chakra-stack css-efzghl"
    >
      <h2
        className="chakra-heading css-kbn7if"
      >
        Please log in
      </h2>
    </div>
  </div>
  <div
    className="css-172bykr"
    spacing="10"
  >
    <div
      className="chakra-stack css-18b10pk"
    >
      <h2
        className="chakra-heading css-odz94x"
      >
        The functionality you want to access requires valid credentials.
      </h2>
      <div
        className="chakra-stack css-1wcf7sk"
      >
        <div
          className="chakra-form-control css-1kxonj9"
          role="group"
        >
          <label
            className="chakra-form__label css-uvho5g"
            htmlFor="username"
            id="username-label"
          >
            Username:
            <span
              aria-hidden={true}
              className="chakra-form__required-indicator css-0"
              role="presentation"
            >
              *
            </span>
          </label>
          <input
            aria-required={true}
            className="chakra-input css-0"
            disabled={false}
            id="username"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            readOnly={false}
            required={true}
          />
        </div>
        <div
          className="chakra-form-control css-1kxonj9"
          role="group"
        >
          <label
            className="chakra-form__label css-uvho5g"
            htmlFor="password"
            id="password-label"
          >
            Password:
            <span
              aria-hidden={true}
              className="chakra-form__required-indicator css-0"
              role="presentation"
            >
              *
            </span>
          </label>
          <div
            className="chakra-input__group css-4302v8"
          >
            <input
              aria-required={true}
              className="chakra-input css-0"
              disabled={false}
              id="password"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Enter password"
              readOnly={false}
              required={true}
              type="password"
            />
            <div
              className="chakra-input__right-element css-fmpuuc"
              size="md"
            >
              <button
                className="chakra-button css-1t02tt6"
                onClick={[Function]}
                type="button"
              >
                Show
              </button>
            </div>
          </div>
        </div>
        <div
          className="css-12p7y6a"
        />
        <button
          className="chakra-button css-oqfhsh"
          onClick={[Function]}
          type="button"
        >
          Log in
        </button>
      </div>
    </div>
  </div>
</div>
`;
